gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala

```diff
@@ -0,0 +1,140 @@
+import org.apache.spark.sql.execution.aggregate.{BaseAggregateExec, HashAggregateExec, ObjectHashAggregateExec}
+
+/**
+ * Merge two phase hash-based aggregate into one aggregate in the spark plan if there is no shuffle:
+ *
+ * Merge HashAggregate(t1.i, SUM, final) + HashAggregate(t1.i, SUM, partial) into
+ * HashAggregate(t1.i, SUM, complete)
+ *
+ * Note: this rule must be applied before the `PullOutPreProject` rule, because the
+ * `PullOutPreProject` rule will modify the attributes in some cases.
+ */
+case class MergeTwoPhasesHashBaseAggregate(session: SparkSession) extends Rule[SparkPlan] {
+
+  val columnarConf: GlutenConfig = GlutenConfig.getConf
+  val scanOnly: Boolean = columnarConf.enableScanOnly
+  val enableColumnarHashAgg: Boolean = !scanOnly && columnarConf.enableColumnarHashAgg
+
+  private def isTransformable(agg: BaseAggregateExec): Boolean = {
+    if (!TransformHints.isAlreadyTagged(agg)) {
+      if (enableColumnarHashAgg) {
+        // Check whether the aggregate is transformable.
+        // Note: do not use the AddTransformHintRule to apply all plan again
+        val transformer = BackendsApiManager.getSparkPlanExecApiInstance
+          .genHashAggregateExecTransformer(
+            agg.requiredChildDistributionExpressions,
+            agg.groupingExpressions,
+            agg.aggregateExpressions,
+            agg.aggregateAttributes,
+            0, // there is no `initialInputBufferOffset` field in Spark 3.2
+            agg.resultExpressions,
+            agg.child
+          )
+        transformer.doValidate().isValid
+      } else false
+    } else {
+      // The transformable tag is already set before
+      TransformHints.isTransformable(agg)
+    }
+  }
+
+  private def isPartialAgg(partialAgg: BaseAggregateExec, finalAgg: BaseAggregateExec): Boolean = {
+    // TODO: now it can not support to merge agg which there are the filters in the aggregate exprs.
+    if (
+      partialAgg.aggregateExpressions.forall(x => x.mode == Partial && x.filter.isEmpty) &&
+      finalAgg.aggregateExpressions.forall(x => x.mode == Final && x.filter.isEmpty)
+    ) {
+      (finalAgg.logicalLink, partialAgg.logicalLink) match {
+        case (Some(agg1), Some(agg2)) => agg1.sameResult(agg2)
+        case _ => false
+      }
+    } else {
+      false
+    }
+  }
+
+  override def apply(plan: SparkPlan): SparkPlan = PhysicalPlanSelector.maybe(session, plan) {
+    if (!BackendsApiManager.getSettings.mergeTwoPhasesHashBaseAggregateIfNeed()) {
+      plan
+    } else {
+      plan.transformDown {
+        case hashAgg @ HashAggregateExec(
+              _,
+              isStreaming,
+              _,
+              _,
+              aggregateExpressions,
+              aggregateAttributes,
+              _,
+              resultExpressions,
+              child: HashAggregateExec)
+            if !isStreaming && isTransformable(hashAgg) && isTransformable(child) =>
+          if (isPartialAgg(child, hashAgg)) {
```

Comment:
why not put this `if` into previous line ?

Reply:
done